### PR TITLE
feature!: impose a minimum loan size when opening a vault

### LINKED ISSUE #4967: Vault creation should have a minimum collateral
Per discussion in wallet meeting, creating a vault should have a governance-controlled minimum collateral parameter at the asset level, which would be enforced only on creation.  

Ideal scenario would be ~$5-$10 range as initial value

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
closes: #4967

## Description

specified by a new governed parameter of the vaultDirector, meaning
the amount is common across all collateral types.

### Security Considerations

Nothing notable. Governance.

### Documentation Considerations

It returns a coherent error message, so no additional documentation necessary.

### Testing Considerations

Had to update many tests, so I could tell the constraint was binding. I added one additional test that fails because of the limit.

### CODE REVIEW

File: packages/run-protocol/src/econ-behaviors.js
Comment (turadg) [inline packages/run-protocol/src/econ-behaviors.js]: Terminology?

File: packages/run-protocol/src/vaultFactory/params.js
Comment (turadg) [inline packages/run-protocol/src/vaultFactory/params.js:52]: Is there any risk in this being a NAT? The brand is implied.

File: packages/run-protocol/src/vaultFactory/vaultDirector.js
Comment (turadg) [inline packages/run-protocol/src/vaultFactory/vaultDirector.js]: ts-expect-error

File: packages/run-protocol/src/vaultFactory/vaultFactory.js
Comment (turadg) [inline packages/run-protocol/src/vaultFactory/vaultFactory.js:59]: 😃

**Comment by Chris-Hibbert** [comment]
@turadg  I improved the typescript in `vaultFactory.js`, getting rid of some `@ts-expect-error` markers, but wasn't able to make `vaultDirector.js` happy, so there are a couple of them there. If you see how to fix it, I'd appreciate an update.

**Review by turadg — APPROVED** [review]
Approving assuming you’ll address comments before merge.

I can fix the types tomorrow.

File: packages/run-protocol/src/vaultFactory/params.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vaultFactory/params.js:52]: It could be a NAT, but I like the clarity of specifying it in RUN. There was discussion in the issue of whether the constraint was on the collateral amount or the loan amount, and specifying it in RUN makes it clear that it's the latter.

File: packages/run-protocol/src/vaultFactory/vaultDirector.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vaultFactory/vaultDirector.js]: done in parallel to your review.

File: packages/run-protocol/src/vaultFactory/params.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vaultFactory/params.js:52]: I ended up changing it in econ-behaviors so it could be defaulted. The brand isn't available during initialization.  This meant the tests had to change to bigints, but the contract still takes an Amount.

File: packages/run-protocol/src/econ-behaviors.js
Comment (dtribble) [inline packages/run-protocol/src/econ-behaviors.js]: Seems like this should at least be a `nat` is not an Amount.

File: packages/run-protocol/src/vaultFactory/params.js
Comment (dtribble) [reply packages/run-protocol/src/vaultFactory/params.js:52]: The debt brand is available here, so I think having this be an Amount is the right thing.

File: packages/run-protocol/src/vaultFactory/vaultDirector.js
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vaultDirector.js]: This seems like it should also be `VaultDirectorParams` above.

File: packages/run-protocol/src/vaultFactory/vaultDirector.js
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vaultDirector.js]: How about just `requestedAmount`?

File: packages/run-protocol/src/vaultFactory/vaultDirector.js
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vaultDirector.js]: `The requested amount (${requestedAmount}) must be at least ${state.directorParamManager.getMinInitialLoan()}`

File: packages/run-protocol/src/vaultFactory/vaultFactory.js
Comment (dtribble) [inline packages/run-protocol/src/vaultFactory/vaultFactory.js]: Don't use "loan" for this.

Comment (dtribble) [reply packages/run-protocol/src/vaultFactory/vaultFactory.js]: An alternative to renaming it is to structure it differently; e.g., "MinCreationFee" and then either always charge at least that, or reject any loan that doesn't have at least that much in fees.

File: packages/run-protocol/test/vaultFactory/test-liquidator.js
Comment (dtribble) [inline packages/run-protocol/test/vaultFactory/test-liquidator.js]: Please use an actual Amount here (like with the initial liquidity below). That was clarifying when I redid the test setup for `test-liquidator`

File: packages/run-protocol/test/vaultFactory/test-liquidator.js
Comment (dtribble) [inline packages/run-protocol/test/vaultFactory/test-liquidator.js]: Please use an actual Amount.

File: packages/run-protocol/test/vaultFactory/test-vaultFactory.js
Comment (dtribble) [inline packages/run-protocol/test/vaultFactory/test-vaultFactory.js]: Why do all the tests have this added. Most should just use the defaults.

**Review by dtribble — CHANGES_REQUESTED** [review]

File: packages/run-protocol/test/vaultFactory/test-vaultFactory.js
Comment (Chris-Hibbert) [reply packages/run-protocol/test/vaultFactory/test-vaultFactory.js]: The production default is on the order of 5_000_000n. I could set the default in the tests to 200n, and it wouldn't effect tests. I could set it to 500n and it would only require a few tests to be updated. Do you have a preference?

File: packages/run-protocol/test/vaultFactory/test-liquidator.js
Comment (Chris-Hibbert) [reply packages/run-protocol/test/vaultFactory/test-liquidator.js]: I used to have an actual Amount. Check the final commit to see.   That didn't work well with econ-behaviors where I needed a default value for some tests in `.../packages/solo/` that didn't set anything.  The default there can't be an amount because the brand isn't available during initialization.

File: packages/run-protocol/src/vaultFactory/vaultFactory.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vaultFactory/vaultFactory.js]: In the issue (#4967) I asked about how to structure the restriction. I think the limit that makes the most sense to users and is straightforward for us to implement is a limit on how much the user borrows. If we need to discuss alternatives further, the context is in that issue.

File: packages/run-protocol/src/econ-behaviors.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/econ-behaviors.js]: fixed

Comment (Chris-Hibbert) [reply packages/run-protocol/src/econ-behaviors.js]: You're right that it should be positive. TypeScript doesn't seem to have `nat`.  I respond about Amounts elsewhere.

File: packages/run-protocol/src/vaultFactory/vaultDirector.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vaultFactory/vaultDirector.js]: tried that.  I'll leave it to @turadg to fix.

File: packages/run-protocol/src/vaultFactory/vaultDirector.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vaultFactory/vaultDirector.js]: good idea.  Done

File: packages/run-protocol/src/vaultFactory/vaultDirector.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vaultFactory/vaultDirector.js]: I went with `The request must be for at least ${getMinInitialDebt}. ${requestedAmount} is too small`

File: packages/run-protocol/test/vaultFactory/test-liquidator.js
Comment (Chris-Hibbert) [reply packages/run-protocol/test/vaultFactory/test-liquidator.js]: responded above.

File: packages/run-protocol/src/vaultFactory/vaultDirector.js
Comment (turadg) [reply packages/run-protocol/src/vaultFactory/vaultDirector.js]: https://github.com/Agoric/agoric-sdk/pull/5243 has the fix. Feel free to click merge on that if you want it in this PR or merge this PR and I'll land the TS fix on master.

File: packages/run-protocol/src/econ-behaviors.js
Comment (dtribble) [inline packages/run-protocol/src/econ-behaviors.js:230]: I'm not sure whether the braces should be closer in all the other lines or whether this needs spaces around braces. Why did the formatting accept this inconsistency?

File: packages/run-protocol/src/econ-behaviors.js
Comment (dtribble) [inline packages/run-protocol/src/econ-behaviors.js:256]: Why is there a hard-coded value here? I thought the default value should come in from governance in the first place.

Comment (dtribble) [inline packages/run-protocol/src/econ-behaviors.js:256]: The centralBrandP is available here so this could just be a real amount?

File: packages/run-protocol/test/vaultFactory/test-liquidator.js
Comment (dtribble) [inline packages/run-protocol/test/vaultFactory/test-liquidator.js]: Note that we have `runBrand` here as well.

Comment (dtribble) [reply packages/run-protocol/test/vaultFactory/test-liquidator.js]: This should set the minimum very low for all the tests that don't care, and then the one that cares should set it to an appropriate value.  `price drop` has nothign to do with this limit and so should not have to mention it.

File: packages/run-protocol/test/vaultFactory/test-liquidator.js
Comment (dtribble) [inline packages/run-protocol/test/vaultFactory/test-liquidator.js]: This test should not have to change this value. Default it to 0n in tests except for the test of the limit.

File: packages/run-protocol/test/vaultFactory/test-vaultFactory.js
Comment (dtribble) [inline packages/run-protocol/test/vaultFactory/test-vaultFactory.js]: As above, this should be 0n in the default test config.

File: packages/run-protocol/test/vaultFactory/test-vaultFactory.js
Comment (dtribble) [inline packages/run-protocol/test/vaultFactory/test-vaultFactory.js:2021]: this should set the min value appropriate for this test.

**Review by dtribble — APPROVED** [review]
The changes in the tests should be fixed, but I don't require a re-review of that. The change between bigint and Amount in the param definition seems like it could still be an amount. But the code looks like it will work fine.

File: packages/run-protocol/src/econ-behaviors.js
Comment (turadg) [reply packages/run-protocol/src/econ-behaviors.js:230]: We don't have formatting of JSDoc yet. Far back burner https://github.com/Agoric/agoric-sdk/pull/4903

fwiw, the tighter is what the (only available jsdoc) autoformatter does. helps distinguish from object braces. e.g. `@param {{ aKey: number, bKey: string }} options`

File: packages/run-protocol/src/econ-behaviors.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/econ-behaviors.js:256]: I was forced into this because some test in `packages/vats` depends on econ-behavior, and doesn't specify any parameters, so the defaults have to work.

Comment (Chris-Hibbert) [reply packages/run-protocol/src/econ-behaviors.js:256]: I tried to do that, but centralBrandP is coming in as a separate parameter, so AFAICT it can't be used to construct a default value for this parameter.

File: packages/run-protocol/test/vaultFactory/test-liquidator.js
Comment (Chris-Hibbert) [reply packages/run-protocol/test/vaultFactory/test-liquidator.js]: done

File: packages/run-protocol/test/vaultFactory/test-liquidator.js
Comment (Chris-Hibbert) [reply packages/run-protocol/test/vaultFactory/test-liquidator.js]: done

File: packages/run-protocol/test/vaultFactory/test-vaultFactory.js
Comment (Chris-Hibbert) [reply packages/run-protocol/test/vaultFactory/test-vaultFactory.js]: done

File: packages/run-protocol/test/vaultFactory/test-vaultFactory.js
Comment (Chris-Hibbert) [reply packages/run-protocol/test/vaultFactory/test-vaultFactory.js:2021]: done

File: packages/run-protocol/src/econ-behaviors.js
Comment (turadg) [reply packages/run-protocol/src/econ-behaviors.js]: it's `NatValue`. that's really `bigint` but it serves as documentation,
https://github.com/Agoric/agoric-sdk/blob/454afd34274017175db5ac0513d6cee5b7375a3a/packages/ERTP/src/types.js#L435

and eventually we could use that information to automate guards.

Comment (Chris-Hibbert) [reply packages/run-protocol/src/econ-behaviors.js]: thanks.  done.

File: packages/run-protocol/src/vaultFactory/vaultDirector.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vaultFactory/vaultDirector.js]: done

**Pull request merged**
